### PR TITLE
Deprecate Python 3.7 due to EOL

### PR DESCRIPTION
**Pull request by jcwchen**
### Description

- Deprecate Python 3.7 due to EOL.
- Use `pip install -e .` instead of `python setup.py install` in CIs, because it is the suggested way to develop onnx mentioned in README.md.
- Set correct permission for clang-tidy.
- Remove dependency: typing_extensions entirely, because starting Python 3.8, native typing support Literal and Protocol.

### Motivation and Context

- Closes https://github.com/onnx/onnx/issues/5163. Python 3.7 will be EOL on 27 Jun 2023.
- Closes https://github.com/onnx/onnx/issues/3750.

### CODE REVIEW

File: onnx/reference/ops/aionnx_preview_training/_op_list.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 import textwrap
 from typing import Any, Dict, Union
+from typing import Optional as TOptional
```
Comment (justinchuby) [inline onnx/reference/ops/aionnx_preview_training/_op_list.py:8]: Why is this alias needed?

**Comment by justinchuby** [comment]
I would also update this line https://github.com/onnx/onnx/blob/a550a45a9da1c7894e63d7a83fc089b02fb82387/pyproject.toml#L64

File: onnx/reference/ops/aionnx_preview_training/_op_list.py (hunk 1 above)
Comment (jcwchen) [reply onnx/reference/ops/aionnx_preview_training/_op_list.py:8]: If we use Optional directly from typing here, there will be some errors:
```
FAILED onnx/test/reference_evaluator_backend_test.py::TestOnnxBackEndWithReferenceEvaluator::test_adagrad - AssertionError: Unable to run test 'test_adagrad' due to Unexpected variable type typing.Optional and class_name='Optional'.
FAILED onnx/test/reference_evaluator_backend_test.py::TestOnnxBackEndWithReferenceEvaluator::test_adagrad_multiple - AssertionError: Unable to run test 'test_adagrad_multiple' due to Unexpected variable type typing.Optional and class_name='Optional'.
FAILED onnx/test/reference_evaluator_backend_test.py::TestOnnxBackEndWithReferenceEvaluator::test_adam - AssertionError: Unable to run test 'test_adam' due to Unexpected variable type typing.Optional and class_name='Optional'.
FAILED onnx/test/reference_evaluator_backend_test.py::TestOnnxBackEndWithReferenceEvaluator::test_adam_multiple - AssertionError: Unable to run test 'test_adam_multiple' due to Unexpected variable type typing.Optional and class_name='Optional'.
FAILED onnx/test/reference_evaluator_backend_test.py::TestOnnxBackEndWithReferenceEvaluator::test_momentum - AssertionError: Unable to run test 'test_momentum' due to Unexpected variable type typing.Optional and class_name='Optional'.
FAILED onnx/test/reference_evaluator_backend_test.py::TestOnnxBackEndWithReferenceEvaluator::test_momentum_multiple - AssertionError: Unable to run test 'test_momentum_multiple' due to Unexpected variable type typing.Optional and class_name='Optional'.
FAILED onnx/test/reference_evaluator_backend_test.py::TestOnnxBackEndWithReferenceEvaluator::test_nesterov_momentum - AssertionError: Unable to run test 'test_nesterov_momentum' due to Unexpected variable type typing.Optional and class_name='Optional'.
```
I guess in some reference code they use Optional from typing as TOptional to distinguish Optional from `onnx.reference.ops.op_optional`.

File: docs/OnnxReleases.md [hunk 2]
```diff
@@ -64,7 +64,7 @@ The ONNX project, going forward, will plan to release roughly on a four month ca
   * Protobuf versions : Latest protobuf version at the time of the release + protobuf version used for previous release
   * Utilize the following matrix to check:
 
-    |   | 3.7 | 3.8 | 3.9 | 3.10 |
+    |   | 3.8 | 3.9 | 3.10 | 3.11 |
```
[the remark this answers is not in the document]
Comment (jcwchen) [reply docs/OnnxReleases.md]: Actually I removed this section in another PR and the reason is described here: https://github.com/onnx/onnx/pull/5208#discussion_r1189191554.

**Review by liqunfu — APPROVED** [review]

**Comment by jcwchen** [comment]
Please note that this PR also removes a dependency: typing_extensions entirely, because starting Python 3.8, native typing support Literal and Protocol.

**Review by gramalingam — APPROVED** [review]

**Pull request merged**